docs/manual/scala/guide/cluster/DomainModelling.md

```diff
@@ -1,19 +1,202 @@
+      // CheckedOut is a final state, not mutations allowed
+      case Checkout(replyTo) =>
+        Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+
+      // it is allowed to read it's state though
+      case Get(replyTo) =>
+        Effect
+        .reply(replyTo)(CartState(items, status = "checked-out"))
+    }
+
+}
+```
+
+Command handlers are the meat of the model. They encode the business rules of your model and act as a guardian of the model consistency. Any mutation must be validated by it. However, they don't apply the mutations. Instead, they express the mutations in the form of persisted events.
+
+Because an Aggregate is intended to model a consistency boundary, it's not recommended to validate commands using data that's not available in scope. Any decision should be solely based on the data passed in the commands and the state of the aggregate. Any external call should be considered a smell because it means that the aggregate is not in full control of the invariants it's supposed to be protecting.
+
+You may have notice that there two ways of sending back a reply. Using `Effect.reply` and `Effect.persist(...).thenReply`. The first one is available directly on `Effect` and should be used when you reply without persisting any event. In this case, you can use the available state in scope because it's guaranteed to not have changed. The second variant should be used when you have persisted one or more events. The updated state is then made available to you on the function used to define the reply.
+
+You may run side-effects inside the command handler. Akka Persistence Typed offers an API for it. Please refer to [Akka documentation](https://doc.akka.io/docs/akka/2.6/typed/persistence.html#effects-and-side-effects) for detailed information.
+
+### Defining the Event Handlers
+
+The event handlers are used to mutate the state of the aggregate by applying the events to it. Event handlers must be pure functions as they will be used when instantiating the aggregate and replaying the event journal. Similar to the command handlers, a [recommended style](https://doc.akka.io/docs/akka/2.6/typed/persistence-style.html#command-handlers-in-the-state) is to add them in the state classes.
+
+```scala
+sealed trait ShoppingCart  {
+  def applyEvent(evt: ShoppingCartEvent): ShoppingCartState
+}
+
+case class OpenShoppingCart(items: Map[String, Int]) extends ShoppingCart {
+
+  def applyEvent(evt: ShoppingCartEvent): ShoppingCartState =
+    evt match {
+      case ItemUpdated(productId, quantity) => updateItem(productId, quantity)
+      case CheckedOut => CheckedOutShoppingCart(items)
+    }
+
+  private def updateItem(productId: String, quantity: Int) =
+    quantity match {
+      case 0 => copy(items = items - productId)
+      case _ => copy(items = items + (productId -> quantity))
+    }
+
+}
+
+// CheckedOut is a final state, there can be any event after its checked out
```

Comment:
```suggestion
// CheckedOut is a final state, there can't be any event after its checked out
```

🤔 Hmmm, I'm not sure what is it you are tr4ying to say. Are you implying any event is valid, or did you actually want to say that there "can't" be any event (as I suggest)?